### PR TITLE
Update registration logic

### PR DESCRIPTION
**Pull request by rudresh**
[#110](https://github.com/DostEducation/RP_IVR_analytics/issues/110)

Added the logic to create registration when first missed call is placed using a phone number.

### Types of changes

- [ ] Bug fix (a change which fixes an issue)
- [x] New feature (a change which adds functionality)

### Checklist:

- [x] I have performed a self-review of my own code.
- [x] The code follows the style guidelines of this project.
- [x] The code changes are passing the CI checks
- [x] I have documented my code wherever required.
- [ ] The changes requires a change to the documentation.
- [ ] I have updated the documentation based on the my changes.

### CODE REVIEW

File: api/services/handle_event_service.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                data["telco_code"] = request.form["TelcoCode"]
+                data["telco_status"] = request.form["TelcoStatus"]
+                data["dial_time"] = request.form["DialTime"]
+                data["end_time"] = request.form["EndTime"]
+                data["duration"] = request.form["Duration"]
+
+                """PickTime is available only when the call is answered.
+                The telco code for answered calls is 16.
+                """
+                if request.form["TelcoCode"] == 16:
```
Comment (Satendra-SR) [inline api/services/handle_event_service.py:36]: Can we get this value from config?

File: api/services/registration_service.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from flask import request
+from api import models, db
+
+
+class RegistrationService:
+    # Create a registration
+    def create_registration(self, data):
+
+        url_decoded_system_phone = requests.utils.unquote(data["to_number"])
+        system_phone_exists = db.session.query(
```
Comment (Satendra-SR) [inline api/services/registration_service.py:12]:
```suggestion
        system_phone = db.session.query(

```

File: api/services/registration_service.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            db.exists().where(models.SystemPhone.phone == url_decoded_system_phone)
+        ).scalar()
+
+        if not system_phone_exists:
+            return
+
+        system_phone_data = models.SystemPhone.query.filter_by(
```
Comment (Satendra-SR) [inline api/services/registration_service.py:19]: Do we need to get these details again? are we not getting these details in line 12?

File: api/services/registration_service.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            phone=url_decoded_system_phone
+        ).first()
+
+        partner = models.PartnerSystemPhone.query.filter_by(
+            system_phone_id=system_phone_data.id
+        ).first()
+
+        user_exists = db.session.query(
```
Comment (Satendra-SR) [inline api/services/registration_service.py:27]: The terminology `user` is a bit confusing here. Let's use registration instead.
```suggestion
        registration_data = db.session.query(

```

File: api/services/registration_service.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            db.exists().where(models.Registration.user_phone == data["from_number"])
```
Comment (Satendra-SR) [inline api/services/registration_service.py:28]: Please check if `data["from_number"]` has the same format we have in the current database.

File: api/models/registration.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    is_child_between_3_6 = db.Column(db.Boolean, nullable=True)
+    is_child_between_3_6 = db.Column(db.Boolean, nullable=True)
+    is_child_above_6 = db.Column(db.Boolean, nullable=True)
+    has_no_child = db.Column(db.Boolean, nullable=True)
+    has_smartphone = db.Column(db.Boolean, nullable=True)
+    has_dropped_missedcall = db.Column(db.Boolean, nullable=True)
+    has_received_callback = db.Column(db.Boolean, nullable=True)
+    status = db.Column(
+        db.Enum(RegistrationStatusEnum),
+        default=RegistrationStatusEnum.PENDING,
```
Comment (Satendra-SR) [inline api/models/registration.py:45]: Enum?

**Review by Satendra-SR — CHANGES_REQUESTED** [review]
@rudresh minor changes. Please have a look

File: api/services/handle_event_service.py (hunk 1 above)
Comment (rudresh) [reply api/services/handle_event_service.py:36]: We have an open issue with picking value from the config in cloud function. I'll make the changes after that.

File: api/services/registration_service.py (hunk 5 above)
Comment (rudresh) [reply api/services/registration_service.py:25]: modified.

File: api/models/registration.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    user_id = db.Column(db.Integer, db.ForeignKey("user.id"))
+    partner_id = db.Column(db.Integer, db.ForeignKey("partner.id"))
+    program_id = db.Column(db.Integer, db.ForeignKey("program.id"))
+    district = db.Column(db.String(100), nullable=True)
+    state = db.Column(db.String(100), nullable=True)
+    parent_type = db.Column(db.String(100), nullable=True)
+    area_type = db.Column(db.String(255), nullable=True)
+    is_child_between_0_3 = db.Column(db.Boolean, nullable=True)
+    is_child_between_3_6 = db.Column(db.Boolean, nullable=True)
+    is_child_between_3_6 = db.Column(db.Boolean, nullable=True)
```
Comment (Satendra-SR) [inline api/models/registration.py:19]: Duplicate.

File: api/services/registration_service.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        if registration_exists:
+            return
+
+        registration = models.Registration(
+            user_phone=data["from_number"],
+            system_phone=url_decoded_system_phone,
+            status="pending",
+            partner_id=partner.id,
+            state=system_phone.state,
+            has_dropped_missedcall=1,
```
Comment (Satendra-SR) [inline api/services/registration_service.py:37]: Thoughts?
```suggestion
            has_dropped_missedcall=True,

```

**Review by Satendra-SR — APPROVED** [review]
Minor changes required. Please fix and merge.

**Pull request merged**